rpcs3/Emu/RSX/Program/GLSLCommon.cpp

```diff
@@ -522,515 +234,221 @@ namespace glsl
-					"\n"
-					"	const bvec2 no_filter = lessThan(abs(uv_step - actual_step), vec2(0.000001));\n"
-					"	if (no_filter.x && no_filter.y)\n"
-					"	{\n"
-					"		return sample0;\n"
-					"	}\n"
-					"\n"
-					"	vec4 a, b;\n"
-					"	float factor;\n"
-					"	const vec4 sample2 = texelFetch2DMS(tex, sample_count, icoords, index, ivec2(0, 1));     // Top left\n"
-					"\n"
-					"	if (no_filter.x)\n"
-					"	{\n"
-					"		// No scaling, 1:1\n"
-					"		a = sample0;\n"
-					"		b = sample2;\n"
-					"	}\n"
-					"	else\n"
-					"	{\n"
-					"		// Filter required, sample more data\n"
-					"		const vec4 sample1 = texelFetch2DMS(tex, sample_count, icoords, index, ivec2(1, 0));     // Bottom right\n"
-					"		const vec4 sample3 = texelFetch2DMS(tex, sample_count, icoords, index, ivec2(1, 1));     // Top right\n"
-					"\n"
-					"		if (actual_step.x > uv_step.x)\n"
-					"		{\n"
-					"		    // Downscale in X, centered\n"
-					"		    const vec3 weights = compute2x2DownsampleWeights(normalized_coords.x, uv_step.x, actual_step.x);\n"
-					"\n"
-					"		    const vec4 sample4 = texelFetch2DMS(tex, sample_count, icoords, index, ivec2(2, 0));    // Further bottom right\n"
-					"		    a = fma(sample0, weights.xxxx, sample1 * weights.y) + (sample4 * weights.z);  // Weighted sum\n"
-					"\n"
-					"		    if (!no_filter.y)\n"
-					"		    {\n"
-					"		        const vec4 sample5 = texelFetch2DMS(tex, sample_count, icoords, index, ivec2(2, 1));    // Further top right\n"
-					"		        b = fma(sample2, weights.xxxx, sample3 * weights.y) + (sample5 * weights.z);  // Weighted sum\n"
-					"		    }\n"
-					"		}\n"
-					"		else if (actual_step.x < uv_step.x)\n"
-					"		{\n"
-					"		    // Upscale in X\n"
-					"		    factor = fract(normalized_coords.x * image_size.x);\n"
-					"		    a = mix(sample0, sample1, factor);\n"
-					"		    b = mix(sample2, sample3, factor);\n"
-					"		}\n"
-					"	}\n"
-					"\n"
-					"	if (no_filter.y)\n"
-					"	{\n"
-					"		// 1:1 no scale\n"
-					"		return a;\n"
-					"	}\n"
-					"	else if (actual_step.y > uv_step.y)\n"
-					"	{\n"
-					"		// Downscale in Y\n"
-					"		const vec3 weights = compute2x2DownsampleWeights(normalized_coords.y, uv_step.y, actual_step.y);\n"
-					"		// We only have 2 rows computed for performance reasons, so combine rows 1 and 2\n"
-					"		return a * weights.x + b * (weights.y + weights.z);\n"
-					"	}\n"
-					"	else if (actual_step.y < uv_step.y)\n"
-					"	{\n"
-					"		// Upscale in Y\n"
-					"		factor = fract(normalized_coords.y * image_size.y);\n"
-					"		return mix(a, b, factor);\n"
-					"	}\n"
-					"}\n\n";
-				};
-
-				insert_msaa_sample_code("sampler2DMS");
+				// Generate multiple versions of the actual sampler code.
+				// We could use defines to generate these, but I don't trust some OpenGL compilers to do the right thing.
+				const std::string_view msaa_sampling_impl =
+					#include "GLSLSnippets/RSXProg/RSXFragmentTextureMSAAOpsInternal.glsl"
```

Comment:
I'm a bit confused how this works xD

Reply:
There's a few rules:
1. an include statement must be on its own line
2. an include statement cannot have a comma on the same line
3. the include directive basically copy-pastes the included file at that point in the code.

So if you create a GLSL file for example and wrap the code with a raw string, you can include it as 
```
variable = // include must be on its own line
#include "myfile.glsl"
; // The previous line will be replaced with a string, so you need to end the "line" with a semicolon.
```